drivers/adc/ad463x/iio_ad463x.c

```diff
@@ -91,31 +151,191 @@ static int32_t _iio_ad463x_prepare_transfer(struct iio_ad463x *desc,
+		return 0;
+	iio_parse_value(buf, IIO_VAL_FRACTIONAL, &vals[0], &vals[1]);
+
+	gain_idx = ad463x_calc_pgia_gain(vals[0], vals[1], ad463x_dev->vref,
+					 ad463x_dev->real_bits_precision);
+	return ad463x_set_pgia_gain(ad463x_dev, gain_idx);
+}
+
+/**
+ * @brief Handles the read request for scale available attribute.
+ * @param dev     - The iio device structure.
+ * @param buf	  - Command buffer to be filled with requested data.
+ * @param len     - Length of the received command buffer in bytes.
+ * @param channel - Command channel info.
+ * @param priv    - Command attribute id.
+ * @return ret    - Result of the reading procedure.
+ *		    In case of success, the size of the read data is returned.
+*/
+static int ad463x_iio_read_scale_avail(void *dev, char *buf,
+				       uint32_t len, const struct iio_ch_info *channel, intptr_t priv)
+{
+	int32_t vals[2];
+	struct iio_ad463x *iio_desc = dev;
+	struct ad463x_dev *ad463x_dev;
+	ad463x_dev = iio_desc->ad463x_desc;
+	char *buffer = buf;
+
+	if(ad463x_dev->has_pgia) {
+		/** Go through the values in the table and add them to the buffer */
+		for (uint8_t i = 0; i < NO_OS_ARRAY_SIZE(ad463x_dev->scale_table); i++) {
+			vals[0] = ad463x_dev->scale_table[i][0];
+			vals[1] = ad463x_dev->scale_table[i][1];
+			buffer += iio_format_value(buffer, len, IIO_VAL_INT_PLUS_NANO, 2, vals) + 1;
+			strcat(buf, " ");
+		}
+		buf[strlen(buf)] = '\0';
+		return strlen(buf);
+	}
+
+	vals[0] = (ad463x_dev->vref * 2) / 1000;
+	vals[1] = ad463x_dev->capture_data_width;
+	return iio_format_value(buf, len, IIO_VAL_FRACTIONAL_LOG2, 2, vals);
+}
+
+/**
+ * @brief Reads the number of given samples for the selected channels
+ * @param dev     - The iio device structure.
+ * @param buf	  - Command buffer to be filled with requested data.
+ * @param samples - Number of samples to be returned
+ * @return ret    - Result of the reading procedure.
+ * 					In case of success, the size of the read data is returned.
+*/
 static int32_t _iio_ad463x_read_dev(struct iio_ad463x *desc, uint32_t *buff,
 				    uint32_t nb_samples)
 {
 	int ret;
-	uint32_t data[nb_samples], i, j, ch;
+	uint16_t total_samples = nb_samples * 2;
+	uint32_t *data;
+	uint32_t i, j, ch;
 
 	if (!desc)
 		return -EINVAL;
 
-	ret = ad463x_read_data(desc->ad463x_desc, data, nb_samples);
-	if(ret)
-		return ret;
+	data = (uint32_t *)no_os_calloc(1, sizeof(uint32_t) * total_samples * 2);
+	if (!data)
+		return -ENOMEM;
+
+	/* Exit register configuration mode */
+	ret = ad463x_exit_reg_cfg_mode(desc->ad463x_desc);
+	if (ret < 0)
+		goto error_comm;
+
+	/** Read samples */
+	ret = ad463x_read_data(desc->ad463x_desc, data, total_samples * 2);
+	if (ret != 0)
+		goto error_comm;
 
-	for(i = 0, j = 0; i < nb_samples; i++)
-		for (ch = 0; ch < desc->iio_dev_desc.num_ch; ch++)
-			if (desc->mask & NO_OS_BIT(ch)) {
-				buff[j++] = data[i];
-			}
+	/** Fill IIO Buffer */
+	for (i = 0, j = 0; i < total_samples; i++)
+		if (desc->mask & NO_OS_BIT(i % 2)) {
```

Comment:
I would avoid `%` operation in a loop, it can slow execution by a lot. Outside a loop nobody would care, but the ineffectivness is multiplied here by total_samples. If you just wanna check if i is a multiple of 2, only check the least significant bit.

Reply:
if (i & 0x1)
    odd
else
    even

Reply:
ok!